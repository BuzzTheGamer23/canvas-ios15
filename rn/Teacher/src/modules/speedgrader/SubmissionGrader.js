// @flow
import React, { Component } from 'react'
import {
  View,
  SegmentedControlIOS,
  StyleSheet,
  Dimensions,
} from 'react-native'
import i18n from 'format-message'
import BottomDrawer from '../../common/components/BottomDrawer'
import GradeTab from './GradeTab'

let { width, height } = Dimensions.get('window')

type State = {
  width: number,
  height: number,
  selectedIndex: number,
}

type SubmissionGraderProps = {
  courseID: string,
  assignmnetID: string,
  userID: string,
  submissionID: ?string,
}

export default class SubmissionGrader extends Component<any, SubmissionGraderProps, State> {
  state: State
  props: SubmissionGraderProps
  drawer: BottomDrawer

  constructor (props: SubmissionGraderProps) {
    super(props)

    this.state = {
      width: width,
      height: height,
      selectedIndex: 0,
    }
  }

  changeTab = (e: any) => {
    this.drawer.open()
    this.setState({
      selectedIndex: e.nativeEvent.selectedSegmentIndex,
    })
  }

  onLayout = (e: any) => {
    this.setState({
      width: e.nativeEvent.layout.width,
      height: e.nativeEvent.layout.height,
    })
  }

  renderTab (tab: ?number): ?React.Element<*> {
    switch (tab) {
      case 0:
        return <GradeTab {...this.props} />
      case 1:
        return <View></View>
      case 2:
        return <View></View>
    }
  }

  render () {
    return (
      <View onLayout={this.onLayout} style={styles.speedGrader}>
        <BottomDrawer ref={e => { this.drawer = e }} containerWidth={this.state.width} containerHeight={this.state.height}>
          <View style={styles.controlWrapper}>
            <SegmentedControlIOS
              testID='speedgrader.segment-control'
              values={[
                i18n({
                  default: 'Grades',
                  description: 'The title of the button to switch to grading a submission',
                }),
                i18n({
                  default: 'Comments',
                  description: 'The title of the button to switch to comments on a submission',
                }),
                i18n({
                  default: 'Files',
                  description: 'The title of the button to switch to the files submitted for a submission',
                }),
              ]}
              selectedIndex={this.state.selectedIndex}
              onChange={this.changeTab}
            />
          </View>
          {this.renderTab(this.state.selectedIndex)}
        </BottomDrawer>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  speedGrader: {
    flex: 1,
  },
  controlWrapper: {
    paddingHorizontal: 16,
    borderBottomWidth: StyleSheet.hairlineWidth,
    borderBottomColor: 'lightgray',
    paddingBottom: 8,
  },
})
