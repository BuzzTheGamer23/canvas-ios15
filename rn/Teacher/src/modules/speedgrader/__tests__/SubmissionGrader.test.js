// @flow

import React from 'react'
import SubmissionGrader from '../SubmissionGrader'
import renderer from 'react-test-renderer'
import DrawerState from '../utils/drawer-state'

jest
  .mock('WebView', () => 'WebView')
  .mock('SegmentedControlIOS', () => 'SegmentedControlIOS')
  .mock('../GradeTab')
  .mock('../components/GradePicker')
  .mock('../components/Header')
  .mock('../components/FilesTab')
  .mock('../comments/CommentsTab')

let template = {
  ...require('../../../api/canvas-api/__templates__/submissions'),
}

let defaultProps = {
  submissionID: '1',
  submissionProps: {},
  drawerState: new DrawerState(),
}

describe('SubmissionGrader', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('renders', () => {
    let tree = renderer.create(
      <SubmissionGrader {...defaultProps} />
    )

    let instance = tree.getInstance()
    let event = {
      nativeEvent: {
        layout: {
          height: 200,
          width: 200,
        },
      },
    }
    instance.onLayout(event)
    // expect(tree.toJSON()).toMatchSnapshot()
  })

  it('can render the handle content', () => {
    let tree = renderer.create(
      <SubmissionGrader {...defaultProps} />
    )
    let handleTree = renderer.create(
      tree.getInstance().renderHandleContent()
    ).toJSON()
    expect(handleTree).toMatchSnapshot()
  })

  it('returns a label with the correct number of files', () => {
    let props = {
      selectedIndex: 0,
      submissionProps: {
        submission: template.submissionHistory([{
          attachments: [{}],
          submission_type: 'online_upload',
        }]),
      },
      drawerState: new DrawerState(),
    }
    let tree = renderer.create(
      <SubmissionGrader {...props} />
    )
    let label = tree.getInstance().filesTabLabel()
    expect(label).toContain('1')
  })

  it('switches between different tabs', () => {
    let tree = renderer.create(
      <SubmissionGrader {...defaultProps} />
    )

    let instance = tree.getInstance()
    let event = {
      nativeEvent: {
        selectedSegmentIndex: 0,
      },
    }
    instance.changeTab(event)
    expect(instance.state.selectedTabIndex).toEqual(0)

    event.nativeEvent.selectedSegmentIndex = 1
    instance.changeTab(event)
    expect(instance.state.selectedTabIndex).toEqual(1)

    event.nativeEvent.selectedSegmentIndex = 2
    instance.changeTab(event)
    expect(instance.state.selectedTabIndex).toEqual(2)
  })

  it('shows number of files on files tab', () => {
    let props = {
      ...defaultProps,
      selectedIndex: null,
      submissionProps: {
        submission: {
          attachments: [
            { fake: 'file' },
            { fake: 'file' },
            { fake: 'file' },
          ],
          submission_type: 'online_upload',
        },
      },
    }

    let tree = renderer.create(
      <SubmissionGrader {...props} />
    )

    let instance = tree.getInstance()
    instance.setState({ height: 200, width: 200, selectedTabIndex: 2 })
    // expect(tree.toJSON()).toMatchSnapshot()
  })

  it('shows no files on files tab', () => {
    let props = {
      ...defaultProps,
      selectedIndex: null,
      submissionProps: {
        submission: {
          fake: 'sub',
        },
      },
    }

    let tree = renderer.create(
      <SubmissionGrader {...props} />
    )

    let instance = tree.getInstance()
    instance.setState({ height: 200, width: 200, selectedTabIndex: 2 })
    tree = tree.toJSON()
    // expect(tree).toMatchSnapshot()
  })

  it('shows number of files on files tab from submission_history', () => {
    let props = {
      ...defaultProps,
      selectedIndex: 1,
      submissionProps: {
        submission: {
          submission_history: [
            { fake: 'sub' },
            {
              attachments: [
                { fake: 'file' },
                { fake: 'file' },
                { fake: 'file' },
              ],
            },
          ],
          submission_type: 'online_upload',
        },
      },
    }

    let tree = renderer.create(
      <SubmissionGrader {...props} />
    )

    let instance = tree.getInstance()
    instance.setState({ height: 200, width: 200, selectedTabIndex: 2 })
    tree = tree.toJSON()
    // expect(tree).toMatchSnapshot()
  })

  it('shows no files on files tab from submission_history', () => {
    let props = {
      ...defaultProps,
      selectedIndex: 0,
      submissionProps: {
        submission: {
          submission_history: [
            {
              fake: 'sub',
            },
          ],
        },
      },
    }

    let tree = renderer.create(
      <SubmissionGrader {...props} />
    )

    let instance = tree.getInstance()
    instance.setState({ height: 200, width: 200, selectedTabIndex: 2 })
    tree = tree.toJSON()
    // expect(tree).toMatchSnapshot()
  })
})
