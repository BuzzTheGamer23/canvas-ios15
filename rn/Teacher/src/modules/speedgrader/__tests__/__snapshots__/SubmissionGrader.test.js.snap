// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmissionGrader can render the handle content 1`] = `
<View
  style={
    Object {
      "borderBottomColor": "lightgray",
      "borderBottomWidth": 0.5,
      "paddingBottom": 8,
      "paddingHorizontal": 16,
    }
  }
>
  <SegmentedControlIOS
    onChange={[Function]}
    selectedIndex={-1}
    testID="speedgrader.segment-control"
    tintColor="#374A59"
    values={
      Array [
        "Grades",
        "Comments",
        "Files",
      ]
    }
  />
</View>
`;
